### DISCUSSION TITLE
Converting from `num::Real` and `Scalar` types to `f32`/`f64`

### DISCUSSION THREAD

**Discussion by okaneco** [Ogeon/palette#341] [Q&A] [answered]
While porting a crate to `palette 0.7`, I had some difficulty figuring out how to convert to a concrete float type. This seems like it'd be hard for beginner-to-intermediate users to figure out.

This was the solution I ended up with for the trait bounds, constraining to `IntoScalarArray<1> + FromScalar<Scalar = f32>`.

```rust
pub fn difference<Wp, T>(c1: palette::Lab<Wp, T>, c2: palette::Lab<Wp, T>) -> f32
where
    T: palette::num::Real
        + palette::num::Arithmetics
        + palette::color_difference::EuclideanDistance
        + palette::num::IntoScalarArray<1>
        + palette::num::FromScalar<Scalar = f32>
        + Copy,
{
    T::into_array(
        palette::color_difference::EuclideanDistance::distance_squared(c1, c2)
    )[0]
}
```

Is there a better way to do this? I was wondering if it made sense to have something similar to `num-traits`'s `ToPrimitive::to_f64` to correspond with the `Real::from_f64` (the return type on `ToPrimitive` is an `Option<f64>`, I don't _think_ it'd be a problem to return `f64`). Or it might just need to be documented somewhere so users can have an easier time finding out how to do this.

**Comment by Ogeon** [comment]
Hmm, I think the issue here is just that you are trying to convert `T` to `f32`. That would depend on what `T` is and Palette doesn't assume much at all (only the operations for `EuclideanDistance` in this case). If you don't need SIMD, you can require it to implement `Into<f32>`, `ToPrimitive` or any other option that fits your use. The traits in `palette::num` shouldn't conflict with `num_traits`, for example, so it's fine to use it in addition. I suppose you could also require `EuclideanDistance<Scalar = f32>`, but that's practically locking `T` to `f32`.

**Reply by Ogeon** [reply]
I'm thinking something like this should work if you also want to convert `T` to `f32` for the output:

```rust
pub fn difference<Wp, T>(c1: palette::Lab<Wp, T>, c2: palette::Lab<Wp, T>) -> f32
where
    T: num_traits::ToPrimitive,
    palette::Lab<Wp, T>: EuclideanDistance<Scalar = T>, // Specifying Scalar may be unnecessary.
{
    c1.distance_squared(c2).to_f32().unwrap() // Or other error handling.
}
```

**Reply by Ogeon** [reply]
You are generally right, though, that writing generic code with Palette is under-documented. It may need to be clarified that it's fine to apply more trait bounds and that the traits in `palette::num` are essentially minimum requirements for maximum generalization. :thinking:

**Reply by okaneco** [reply]
I appreciate the answers.

The interface designs are from when I was somewhat new to Rust so there are things I'd like to change; I wanted to make it easier for users and I wasn't the most confident with generics at the time. I was implicitly relying on `Lab<Wp>` since `palette 0.5` which masked some of the issues since it defaulted to `f32`. Upgrading to `0.6` was uneventful, but `0.7` exposed the mistaken assumptions. I ended up with the code above because but I wanted to upgrade to `0.7` without majorly changing the public interface. The only user facing changes they should have to make are converting from `Pixel::{into_raw, from_raw_slice, into_raw_slice}` to the `cast` equivalents which I can easily cover in the change log.

An issue was filed https://github.com/okaneco/kmeans-colors/issues/51 and it wasn't obvious how to upgrade to the new version because of the flawed assumptions I'd already made (such as implicitly relying on `Copy` and `PartialOrd` in some places). I was able to work on it again after porting a different crate with a fresh set of eyes. This gave me a better idea of what I was doing, and brought me to an understanding that echoes your words.

> It may need to be clarified that it's fine to apply more trait bounds and that the traits in palette::num are essentially minimum requirements for maximum generalization.

I tried to make everything work with only the `palette::num` traits at first, but soon realized that wouldn't be possible with what was available. That was a faulty assumption on my part since I was under the impression it was still a wrapper around `Float`, and I was hesitant to add more traits.

Do you have a recommendation of when users should prefer `num-traits` versus `palette::num`? Reading the documentation more, it seems like unless you're using `wide`, you should probably be using the `num-traits` instead. Making that more clear or with cookbook examples would be helpful here.

**Reply by Ogeon** [reply, marked as answer]
I see, sorry for the headaches. First, let me just answer your last question: prefer `num-traits` over `palette::num` as much as you possibly can for now. It's much more complete and long-term stable, and also more integrated in the general ecosystem. `palette::num` should ideally be a separate crate but I didn't want to take that step without trying it out in practice first (and I'm not sure I should be the one maintaining it, honestly).

For the case with `Lab` (I found your PR and had a look), since you effectively required `Lab<Wp, f32>` before, you should be able to continue doing that and not make it generic over the component type for now. You shouldn't have to introduce new requirements in that case, but it may be wise to make the `f32` explicit. I'm sometimes wondering if it was a mistake to make the component type a default, to be honest.

But if you do want to make it fully generic, I would suggest sticking to `num-traits`, as mentioned, and use as many high level `palette` trait bounds as possible, rather than restricting the type `T` to make them work. So `Lab<Wp, T>: FromColor<LinSrgb<f32>>`, `Lab<Wp, T>: EuclideanDistance`,  and similar. It's easier to maintain and read in general. Think of `palette::num` as an implementation detail. You would also allow a larger set of types `T`  and any bounds on it from `palette` should be valid for `f32` too. What would happen, though, is that you start exposing details, such as which distance formula you are using, but that's the unfortunate nature of generic math code in Rust. 😬 It's hard to know where to put the requirements.

I hope this helped putting you back on track, but let me know if not. I'm happy to help where I can and I could also comment with suggestions in the PR if you want.

**Reply by okaneco** [reply]
Thanks, that does help put me back on course and give me a direction to aim towards. I'll try to re-approach it with `Lab<Wp, f32>` and/or `num-traits` if necessary. If I have more questions I'll be sure to post here or ping for comments at that PR.

> What would happen, though, is that you start exposing details, such as which distance formula you are using, but that's the unfortunate nature of generic math code in Rust. 😬 It's hard to know where to put the requirements.

Yes, a constant struggle 😅

**Reply by okaneco** [reply]
I went the `num-traits` route and was able to finish with relative ease. The main files of interest are `src/colors/kmeans.rs` and `src/colors/sort.rs`.  
https://github.com/okaneco/kmeans-colors/pull/52/commits/e61e081907531d317fbc612b1345b9a0a086bfc6

I decided to go generic over the component type and these were the bounds I arrived at. I also changed from `Srgb` to `Rgb` to get away from the implied/default values of the color type and the sometimes weird inference errors because of `Srgb` being an alias.
```rust
impl<Wp, T> Calculate for Lab<Wp, T>
where
    T: Float + FromPrimitive + Zero,
    Lab<Wp, T>: core::ops::AddAssign<Lab<Wp, T>> + Default,
```
```rust
impl<S, T> Calculate for Rgb<S, T>
where
    T: Float + FromPrimitive + Zero,
    Rgb<S, T>: core::ops::AddAssign<Rgb<S, T>> + Default,
```

The reminder about using higher level bounds was very helpful because it allowed me to remove unnecessary implementation traits. I originally had `S: RgbStandard` and `T: IntoStimulus<f32>` but they were replaced by an `IntoColor<Luma<S, T>>` bound for a function that sorted values by brightness after converting to `Luma`.
```rust
impl<S, T> Sort for Rgb<S, T>
where
    T: Float + FromPrimitive + Zero,
    Rgb<S, T>: core::ops::AddAssign<Rgb<S, T>> + IntoColor<Luma<S, T>> + Default,
```

Thanks for helping me resolve the confusion I originally had 🙂

**Reply by Ogeon** [reply]
Nice! Looks like you found ways to simplify some of the code, too. :tada: It makes sense to generalize `Rgb` as well, since `Lab` takes basically anything, and it should allow a wider range of image formats too.

If I may come with a small suggestion, you could check if `into_array` can be replaced with `.into()`. It should work if the type inference understands it. Unless you prefer `into_array`, of course! The other cast functions will also have trait/method counterparts in an upcoming version ([implemented as `ComponentsInto` etc.](https://ogeon.github.io/palette/palette/cast/trait.ComponentsInto.html)), but I have some more I want to add first so don't wait.

> The reminder about using higher level bounds was very helpful because it allowed me to remove unnecessary implementation traits.

That's great! I have become dangerously fond of it, myself, since it's much more "semantic" and usually makes the intentions quite clear. It's just too bad that the error messages still tend to dip down into the detailed bounds of the implementations. Maybe it's hard to know which level to put it on. I know it's possible to customize the error message on nightly, but that's not stable...

**Reply by okaneco** [reply]
Thanks, I was able to remove `into_array` and simplify with `into()`; in another crate, I was able to remove `into_array_ref` and use `into()`. I often forget about using `into()` because of inference not always working.

`ComponentsInto` looks like a really nice convenience feature, I look forward to being able to use it.

**Reply by Ogeon** [reply]
I wasn't sure if it had all the required context in your case, but nice to see that it did!

> `ComponentsInto` looks like a really nice convenience feature, I look forward to being able to use it.

I'll try not to take forever with it, then! :laughing: I just want to let it sit for a bit before I push it out (and also finish CAM16). I was deliberately avoiding anything like `ComponentsInto` and friends in the beginning, to let the "low level" API settle a bit, but it feels more clear how to make something convenient on top now. I'm also thinking of adding counterparts to `AsRef` and `AsMut`, essentially to always borrow, so that should be included as well. So instead of having to write `(&mut buffer).components_into()`, it could be `buffer.components_as_mut()`.

**Reply by okaneco** [reply]
No rush 😃 

The CAM16 work looks exciting. I tried working on CAM02 conversion a long time ago but it was very clumsy trying to handle all the parameters needed to perform the conversion. It seemed like it could be a good place for the pipelining idea that you mentioned a while ago #197. I lost track of whether that made it into the library or not but it seemed helpful in some heavier scenarios of batch operations on colors such as repeatedly passing the parameter pack for color conversion.

**Reply by Ogeon** [reply]
I never got around to it, but it's still on the table for the future. Most people don't need it (I can put and have put in targeted optimizations for the sRGB/D65 cases) but I think it's still useful for the "power users", who would be interested in customizing things. Speaking of optimizations, I forgot to mention that you _may_ get a few % out of using `.into_linear::<f32>()` instead of `.into_format()` on your `Srgb<u8>` values, if that stage is significant enough. It uses a look-up table, so it is faster, but I don't know if it's even noticeable in your case. I will try to make it part of `.{from,into}_color()` in 0.8, when I can afford messing up the type inference for people.

But yeah, CAM16 is complex in the same way as I imagine CAM02 was and I have to figure out a good balance between static and dynamic parameters. I'm experimenting with `BakedParameters` as an option, so it's possible to cache the part of it that's independent of the color. My main head scratcher right now is how to deal with partial information. The current `PartialCam16` in the PR isn't great when I add in CAM16-UCS, since there's no longer a universally "best" set of partial information. :thinking: I can see how this task could have too many "design questions" for someone who isn't already neck deep in this library. :eyes: I can go on way too long about how I wish I could do it and the obstacles so far, but that's probably better as a topic for the PR or the issue. I can foresee some breaking changes in that area in 0.8. :grimacing:

**Reply by okaneco** [reply]
### Comparing `into_format` and `into_linear`

> Speaking of optimizations, I forgot to mention that you may get a few % out of using `.into_linear::<f32>()` instead of `.into_format()` on your `Srgb<u8>` values, if that stage is significant enough. It uses a look-up table, so it is faster, but I don't know if it's even noticeable in your case.

Ah, so that's where `fast_srgb8` is used. I need to set up a more proper benchmarking harness, but it seems to dramatically improve performance 🎉

This does the "minimal" amount of real-world work going one-way decoding a JPEG image, converting from `Rgb<u8>` to `Lab<D65, f32>`, solving for 8 centroid colors and the corresponding indices, then exiting without saving an image file. Ran a handful of times with little variance.
```bash
time target/release/kmeans_colors -i FILE --no-file
```
 
|Image Size|`into_format`|`into_linear`|
|:-:|-:|-:|
|4547x3032|7.8s|**4.3s**|
|1700x2200|2.4s|**1.3s**|
|805x1000|0.48s|**0.35s**|
|320x206|0.056s|**0.046s**|

I was happy with just the 15-17% improvement I saw on the small images, but the large image conversions are a huge win on my machine. A _lot_ more time was spent in that conversion than I expected.

Unfortunately for a lot of the projects I'm interested in, the wins are going to be one-directional since they're more quantization related and will only need 2-256 color conversions back from `Lab` to `Rgb`. But I imagine other workflows may greatly benefit going from `Lab` to `LinSrgb` to `Srgb::from_linear`.

#### Other potential hot spots

When I ran a flamegraph maybe 2 years ago, the functions that I believe showed up by name were the `Rgb`<->`Xyz` matrix conversions (or functions that called into the 3x3 multiply and inverse) and `Lab::from_color_unclamped(Xyz)` (cube root?) but I'm not 100% confident on that. I'm not sure if some caching/batch conversion could make the `Rgb`/`Xyz` parts faster or if it's already being as reasonably optimized as it can be.

---

>  I can see how this task could have too many "design questions" for someone who isn't already neck deep in this library.

Yes, as you know it's a lot (CAM02 also has a UCS extension which was difficult to think about in terms of the integration). It seemed like something that would demand a large rewrite of the library or for the library to be designed around so I figured I'd leave that task to the experts. 😄

**Reply by Ogeon** [reply]
Oh, that's fantastic! :tada: I wasn't sure, as you saw, if it would be significant enough, but almost cutting the time in half for large images is just incredible! The `from_linear` function should also be faster with `fast_srgb8` integrated if it goes straight from `f32` or `f64` to `u8`, so I think your other use cases should also have gotten a boost if they fit that pattern. That's also the main reason why I picked a crate for it, since that process is more complex than a simple table lookup.

The `Rgb` <-> `Xyz` conversion should be faster as well. It has hard coded matrices for the sRGB primaries, so it shouldn't have to invert the matrix anymore. Other primary sets can choose do the same. It's possible that it could be optimized further, but that was the obvious low hanging fruit.

I haven't looked too closely at what to do with `Lab` yet, other than confirming that it's comparatively expensive. Going the other way (`Lab` to `Xyz`) shows up as much faster in my PR benchmarks, so it's possibly something like the cube root that's slowing it down. I think this is an opportunity for caching if your input is an RGB8 image, unless you already do that. :thinking: I saw a similar project use the opaque colors as keys in a `HashMap` to memoize the results. It could work if the hashing and lookup is faster, sacrificing some more memory instead.

> Yes, as you know it's a lot (CAM02 also has a UCS extension which was difficult to think about in terms of the integration). It seemed like something that would demand a large rewrite of the library or for the library to be designed around so I figured I'd leave that task to the experts. :smile:

That's where I am right now. It's getting a bit wild but the system is still holding up, amazingly. It hinges on having a default value for the viewing conditions, though. The alternative would be to break the chain of conversions at the `Xyz`/`Cam16` connection, so the parameters can be provided. Then it's straight forward to convert to UCS, and back to a partial CAM16. After that, the parameters are needed to reconstruct the other values.

**Reply by okaneco** [reply]
_This comment is a summary of caching color conversions, no questions in it_

I tried caching `Srgba<u8>` to `Lab<f32>` conversions, and it seems very much worth doing.
[Here's the gist](https://gist.github.com/okaneco/f57314d37792462626fe9ac4fe2eff21)

|4547x3032 pixels|conversions|time|improvement|
|:-:|-:|-:|:-:|
|`into_format`|13786504|3.87s||
|`into_linear`|13786504|1.73s|2.2x|
|`into_linear` + caching|118038|213ms|18.2x|

|1700x2200 pixels|conversions|time|improvement|
|:-:|-:|-:|:-:|
|`into_format`|3740000|1.14s||
|`into_linear`|3740000|476ms|2.4x|
|`into_linear` + caching|91508|57.2ms|19.9x|

|805x1000 pixels|conversions|time|improvement|
|:-:|-:|-:|:-:|
|`into_format`|805000|236ms||
|`into_linear`|805000|107ms|2.2x|
|`into_linear` + caching|74166|23ms|10.3x|

|320x206 pixels|conversions|time|improvement|
|:-:|-:|-:|:-:|
|`into_format`|65920|19.4ms||
|`into_linear`|65920|8.6ms|2.3x|
|`into_linear` + caching|19910|3.9ms|4.9x|

The simplest version of inserting into the map while iterating ended up being the fastest.[^1]

```rust
fn cached_rgb_to_lab(
    rgb: &[Srgba<u8>],
    lab: &mut Vec<Lab<D65, f32>>,
    map: &mut FxHashMap<[u8; 3], Lab<D65, f32>>,
) {
    lab.extend(rgb.iter().map(|color| {
        *map.entry([color.red, color.green, color.blue])
            .or_insert_with(|| color.into_linear::<_, f32>().into_color())
    }))
}
```

I assumed that memoization was going to be slower due to hashing/lookup/reallocation, but that it might have some inflection point where it became faster as image size increased. That turned out not to be the case for general images I've tried.

A worst case scenario of every RGB color still ended up being faster than the current `into_format` pipeline
http://www.brucelindbloom.com/index.html?RGB16Million.html

|4096x4096 pixels|time|conversions|
|:-:|:-:|:-:|
|`into_format`|4.81s|16777216|
|`into_linear`|2.29s|16777216|
|`into_linear` + caching|4.44s|16777216|

[^1]: I tried two other methods but runtime seemed to be dominated by reading the image twice. The first approach populated the map with `Lab::default()`, then iterated mutably through all the KV pairs in the map calculating the correct value, then read through the image again looking up the colors and inserting into the `Lab` vec. The second approach populated the map with the converted `Lab` color in the first image read, but didn't insert to the output vector until reading through the image again. The second approach was faster than the first approach. Both approaches were faster than no use of caching but not as fast as the simplest approach above.

**Reply by okaneco** [reply]
_Putting the question up front:_
Should there be `IntoLinear<T, U>` or `IntoLinear<T, u8>` impls for `Srgb`?

While writing out that benchmark program, I tried to write some functions generically and wasn't able to for the ones that use `into_linear`. I was able to write the first.

```rust
fn regular<Wp, T, U>(rgb: &[Srgba<T>], lab: &mut Vec<Lab<Wp, U>>)
where
    Lab<Wp, U>: FromColor<Srgba<U>>,
    T: Copy,
    U: FromStimulus<T>,
{
    lab.extend(rgb.iter().map(|&x| x.into_format().into_color())) // compiles
}
```

I imagine the ones that use `into_linear` might look something like this but it doesn't compile. It only picks up the `IntoLinear<T, T>` or the explicit `f32`/`f64` as suggestions, and I get this error under `into_linear`.


```rust
fn linear<Wp, T>(rgb: &[Srgba<u8>], lab: &mut Vec<Lab<Wp, T>>)
where
    Lab<Wp, T>: FromColor<LinSrgba<T>>,
    LinSrgb<T>: FromColor<Srgb<u8>>,
    T: FromStimulus<u8>,
{
    lab.extend(rgb.iter().map(|x| x.into_linear().into_color())) // doesn't compile
}
```
```
the trait bound `Srgb: IntoLinear<T, u8>` is not satisfied
the trait `IntoLinear<T, u8>` is not implemented for `Srgb`
```
---

I can write out the code I need to explicitly with no problems. There just seemed to be some pain points around type inference with `LinSrgb` that I've noticed. Writing code generic over the white point looks like this. I'd love to be able to write the second but I understand there are limits sometimes with traits/annotations.
```rust
fn linear<Wp>(rgb: &[Srgba<u8>], lab: &mut Vec<Lab<Wp, f32>>)
where
    Lab<Wp, f32>: FromColor<LinSrgba<f32>>,
    Wp: WhitePoint<f32>,
{
    lab.extend(
        rgb.iter()
            .map(|x| Lab::<Wp, f32>::from_color(x.into_linear())), // compiles
        // .map(|x| x.into_linear().into_color()), // doesn't compile
    )
}
```

**Reply by Ogeon** [reply]
*If you get double notifications for this, it's because I posted it outside the thread at first. Oops!*

Awesome improvements! 🎉 I think the simplest version is what I would have gone for as well. That way I think you only pay for what you use. In your first approach, you pay for the memory space and hashing in advance and then use an unknown portion of it in practice. In the second approach you hash and look up twice, while in the simplest approach you only hash and look up once. The entry API is really clever in that way! Oh, and another small tip here is that you can use `color.into()` for converting to an array, but the explicit style is nice in its own way.

----

For making it more generic, it's in practice only available for sRGB out of the box, but third party crates can always add more options. As you figured out, it doesn't work with just `FromColor`, and this is because I didn't add the component type conversion as part of the implementation. I simply forgot. 😅 Adding it is going to mess up type inference, so I prefer to wait with it until 0.8. It's also not related to `FromStimulus`, which does 1:1 conversions within the full ranges of unsigned integers or within 0.0-0.1 for real-ish numbers.

What you can do is to require the RGB standard's transfer function to implement `IntoLinear<T, u8>`. You can also make it generic over RGB standards at the same time, but that's going to be mostly sRGB in practice, as mentioned above. So, for example, in its simplest form:

```rust
fn linear<Wp, T>(rgb: &[Srgba<u8>], lab: &mut Vec<Lab<Wp, T>>)
where
    Lab<Wp, T>: FromColor<LinSrgba<T>>, // This is probably only working for `Wp = D65`, BTW
    palette::encoding::Srgb: IntoLinear<T, u8>,
{
    lab.extend(rgb.iter().map(|x| x.into_linear().into_color()))
}
```

You can check the [`into_linear`](https://docs.rs/palette/0.7.2/palette/rgb/struct.Rgb.html#method.into_linear) method for that bound, but in its most generic form. `Srgb` has itself as both transfer function and RGB space. If you want to support any standard that has this specific option, you will have to deal with some extra traits and associated types. 😬 It would be something like this:

```rust
fn linear<S, Wp, T>(rgb: &[Rgba<S, u8>], lab: &mut Vec<Lab<Wp, T>>)
where
    Lab<Wp, T>: FromColor<Rgba<Linear<S::Space>, T>>,
    S: RgbStandard,
    S::TransferFn: IntoLinear<T, u8>,
{
    lab.extend(rgb.iter().map(|x| x.into_linear().into_color()))
}
```

It gets a bit hairy from that point. As a side note, I would like to see if I can improve that too, at some point, (less trait nesting, at least) but that's also a future issue.

I'm not sure why your last example doesn't compile. It should work unconditionally, so it's possibly something with the trait resolution. hard to say without seeing the errors.

**Reply by okaneco** [reply]
_I drafted this the day after your response and forgot to post it. Cleaned it up a little, hope it's not too disorganized._

Thanks for the reminder again on `color.into()`, I keep forgetting that. For the hashing, I just wanted to cover all the bases in case something surprising were to happen. But as you said, the entry API is really the star.

---

> For making it more generic, it's in practice only available for sRGB out of the box, but third party crates can always add more options. As you figured out, it doesn't work with just `FromColor`, and this is because I didn't add the component type conversion as part of the implementation. I simply forgot. 😅 Adding it is going to mess up type inference, so I prefer to wait with it until 0.8. It's also not related to `FromStimulus`, which does 1:1 conversions within the full ranges of unsigned integers or within 0.0-0.1 for real-ish numbers.

Yes, for that I didn't mind using `Srgba` because I mainly wanted to make it generic over the `Lab` parameters and didn't plan on using any other transfer functions. I think I can see now how it'd mess up type inference so I understand waiting for the next breaking version, and also that it's probably only valid for the `D65` white point.

Those code snippets were very helpful. I get the same compiler error which for both which goes away after adding an `u8: IntoStimulus<T>` or `T: FromStimulus<u8>` bound, as follows.

```rust
error[E0277]: the trait bound `u8: IntoStimulus<T>` is not satisfied
   --> examples\cache.rs:92:37
    |
92  |     lab.extend(rgb.iter().map(|x| x.into_linear().into_color()))
    |                                     ^^^^^^^^^^^ the trait `IntoStimulus<T>` is not implemented for `u8`
    |
    = note: required for `T` to implement `FromStimulus<u8>`
note: required by a bound in `rgb::rgb::<impl Alpha<palette::rgb::Rgb<S, T>, A>>::into_linear`
   --> .cargo\registry\src\index.crates.io-6f17d22bba15001f\palette-0.7.2\src\rgb\rgb.rs:482:12
    |
479 |     pub fn into_linear<U, B>(self) -> Alpha<Rgb<Linear<S::Space>, U>, B>
    |            ----------- required by a bound in this associated function
...
482 |         B: FromStimulus<A>,
    |            ^^^^^^^^^^^^^^^ required by this bound in `rgb::rgb::<impl Alpha<Rgb<S, T>, A>>::into_linear`
help: consider extending the `where` clause, but there might be an alternative better way to express this requirement
    |
89  |     palette::encoding::Srgb: IntoLinear<T, u8>, u8: IntoStimulus<T>
    |                                               ~~~~~~~~~~~~~~~~~~~~~
```
```rust
fn linear<S, Wp, T>(rgb: &[Rgba<S, u8>], lab: &mut Vec<Lab<Wp, T>>)
where
    Lab<Wp, T>: FromColor<Rgba<Linear<S::Space>, T>>,
    S: RgbStandard,
    S::TransferFn: IntoLinear<T, u8>,
    u8: IntoStimulus<T>,
```

---

Sorry for not including the errors in the last example I posted here https://github.com/Ogeon/palette/discussions/341#discussioncomment-6647996

```rust
error[E0282]: type annotations needed
   --> examples\cache.rs:108:51
    |
108 |     lab.extend(rgb.iter().map(|x| x.into_linear().into_color()))
    |                                                   ^^^^^^^^^^
    |
help: try using a fully qualified path to specify the expected types
    |
108 |     lab.extend(rgb.iter().map(|x| <Alpha<palette::rgb::Rgb<Linear<Srgb>, U>, B> as IntoColor<T>>::into_color(x.into_linear())))
    |                                   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++               ~

error[E0283]: type annotations needed
   --> examples\cache.rs:108:37
    |
108 |     lab.extend(rgb.iter().map(|x| x.into_linear().into_color()))
    |                                     ^^^^^^^^^^^ cannot infer type of the type parameter `U` declared on the method `into_linear`
    |
    = note: multiple `impl`s satisfying `Srgb: IntoLinear<_, u8>` found in the `palette` crate:
            - impl IntoLinear<f32, u8> for Srgb;
            - impl IntoLinear<f64, u8> for Srgb;
note: required by a bound in `rgb::rgb::<impl Alpha<palette::rgb::Rgb<S, T>, A>>::into_linear`
   --> .cargo\registry\src\index.crates.io-6f17d22bba15001f\palette-0.7.2\src\rgb\rgb.rs:481:24
    |
479 |     pub fn into_linear<U, B>(self) -> Alpha<Rgb<Linear<S::Space>, U>, B>
    |            ----------- required by a bound in this associated function
480 |     where
481 |         S::TransferFn: IntoLinear<U, T>,
    |                        ^^^^^^^^^^^^^^^^ required by this bound in `rgb::rgb::<impl Alpha<Rgb<S, T>, A>>::into_linear`
help: consider specifying the generic arguments
    |
108 |     lab.extend(rgb.iter().map(|x| x.into_linear::<U, B>().into_color()))
    |                                                ++++++++

error[E0283]: type annotations needed
   --> examples\cache.rs:108:51
    |
108 |     lab.extend(rgb.iter().map(|x| x.into_linear().into_color()))
    |                                                   ^^^^^^^^^^
    |
note: multiple `impl`s or `where` clauses satisfying `_: FromColor<Alpha<palette::rgb::Rgb<Linear<Srgb>, _>, _>>` found
   --> examples\cache.rs:105:19
    |
105 |     Lab<Wp, f32>: FromColor<LinSrgba<f32>>,
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^
    = note: and another `impl` found in the `palette` crate:
            - impl<T, U> FromColor<T> for U
              where U: FromColorUnclamped<T>, U: palette::Clamp;
    = note: required for `Alpha<palette::rgb::Rgb<Linear<Srgb>>, f32>` to implement `IntoColor<Lab<Wp>>`
help: try using a fully qualified path to specify the expected types
    |
108 |     lab.extend(rgb.iter().map(|x| <Alpha<palette::rgb::Rgb<Linear<Srgb>, U>, B> as IntoColor<T>>::into_color(x.into_linear())))
    |                                   +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++               ~
```

**Reply by Ogeon** [reply]
> Those code snippets were very helpful. I get the same compiler error which for both which goes away after adding an u8: IntoStimulus<T> or T: FromStimulus<u8> bound, as follows.

Right, I forgot about the alpha channel. It's always linear, so it needs its own mapping.

> Sorry for not including the errors in the last example I posted here #341 (reply in thread)

Looks like it's having trouble inferring the linear component type through those two steps. I suppose you give it enough information in the working case to figure out how to make it linear too. 🤔 I'm still surprised that it doesn't see what the end result should be and infers it from there...
